# 응답

### 앞에서 `nest g resource Users` 명령어로 Users 리소스에 대한 CRUD API를 만들어 보았다. 서버를 실행하면 어떤 라우팅 패스를 통해 요청을 받을 수 있는지 콘솔 로그를 통해 확인할 수 있다.

<img width="793" alt="스크린샷 2022-04-26 오후 3 19 30" src="https://user-images.githubusercontent.com/25292654/165234615-89cbd964-ba9c-4dfd-a2dc-400b8fafbc05.png">

이를 참고하여 각각 요청해보면 다음과 같은 결과가 나온다.


| 경로 | http method | 응답 상태 코드 | body |
| --- | --- | --- | --- |
| /users | POST	| 201 |	This action adds a new user |
| /users | GET | 200 | This action returns all users |
| /users/1 | GET | 200 | This action returns a #1 user |
| /users/1 | PATCH | 200 | This action updates a #1 user |
| /users/1 | DELETE |	200 |	This action removes a #1 user |

### ※ CLI로 자동 생성된 update는 PATCH 메서드를 사용하고 있다. http 메서드에는 업데이트 동작을 기술하는 메서드가 2가지 있는데 PUT은 리소스전 체를 교체할 때 쓰고, PATCH는 리소스의 일부를 업데이트 할 때 사용한다. 하지만 실제 구현시에는 이를 엄격하게 지키지 않고 PUT을 보통 사용하는데 만약 PATCH가 사용되었다면 이같은 뜻을 가진다고 생각하면 된다.

- 각 요청의 응답 코드는 POST일 경우엔 201, 나머지는 200이다. 또한 응답 본문은 스트링 값을 가지고 있는데 이것은 UsersController의 각 메서드가 리턴하는 값이다.

- Nest는 이렇게 응답을 어떤 방식으로 처리할것인지 미리 정의해 두었다. string, number, boolean과 같이 자바스크립트 원시 타입을 리턴할 경우 직렬화 없이 바로 보내지만, 객체를 리턴할 땐 직렬화를 톧해 JSON으로 자동 변환해 준다. 이 방법이 권장하는 방법이지만 라이브러리별 응답 객체를 직접 다룰 수도 있다. 예를 들어 Express를 사용한다면 [Express response object](https://expressjs.com/en/api.html#res)를 `@Res 데코레이터를 이용해서 다룰수도 있다.

```typescript
@Get()
findAll(@Res() res) {
  const users = this.usersService.findAll()

  return res.status(200).send(users);
}
```

앞서 Nest는 CRUD에 대한 성공 응답으로 POST는 201, 나머지는 200을 가진다고 했는데 만약 이 상태 코드를 다른 값으로 바꾸고 싶다면 데코레이터 `@HttpCode`를 사용하면 된다.

```typescript
import { HttpCode } from '@nestjs/common';

@HttpCode(202)
@Patch(':id')
update(@Param('id') id: string, @Body() updateUserDto: UpdateUserDto) {
  return this.usersService.update(+id, updateUserDto);
}
```

### ※ HTTP 202 Accepted, 요청이 성공적으로 접수되었지만 아직 해당 요청에 대해 처리중이거나 처리 시작 전임을 의미한다. 불행히도 요청이 처리중 실패할 수도 있기 때문에 요청은 실행될수도 되지 않을수도 있다. 이 상태 코드는 비확약적, 즉 HTTP가 나중에 요청 처리 결과를 나타내는 비동기 응답을 보낼 방법이 없다는 것을 의미한다. 이는 다른 프로세스나 서버가 요청을 처리하는 경우 또는 일괄 처리를 위한 것이다. - `MDN Web Docs`

- 만약 요청을 처리하는 도중 에러가 발생했거나 예외를 던져야 한다면 어떻게 해야 할까? 

- 예를 들어 보자, 유저 정보 조회(GET /users/:id)를 요청했는데 id는 1부터 시작하는 규칙을 가지고 있다고 가정해보자. 만약 id가 1보다 작은 값이 전달 될 경우엔 400 Bad Request 예외를 발생시켜야한다.

```typescript
@Get(':id')
findOne(@Param('id') id: string) {
  if (+id < 1) {
    throw new BadRequestException('id는 0보다 큰 값이어야 합니다.');
  }

  return this.usersService.findOne(+id);
}
```

예외를 던져주는 코드를 작성한 뒤 id를 0으로 요청하면 다음과 같이 나온다.

```
$ curl -X GET http://localhost:3000/users/0
{
  "statusCode": 400,
  "message": "id는 0보다 큰 값이어야 합니다.",
  "error": "Bad Request"
}
```

NotFoundException 객체의 생성자로 전달한 메세지와 함께 상태코드가 400인 에러가 발생한다.

※ 터미널에서 curl 명령어로 실행하면 위와 같이 JSON 결과가 읽기 편하게 출력되지 않는다. 출력을 jq에 파이프로 전달하는 방법을 찾아보자.
  - ex) <code> $ curl -X GET http://localhost:3000/users/0 | jq </code>